doc/development/autograph.rst

```diff
@@ -0,0 +1,743 @@
+Guide for AutoGraph for plxpr capture
+=====================================
+
+When capturing PennyLane programs as a plxpr instance using AutoGraph, you
+can represent quantum programs with **structure**. That is, you can use
+classical control flow (such as conditionals and loops) with quantum operations
+and measurements, and this structure is captured and preserved in the plxpr
+representation.
+
+PennyLane provides various high-level functions, such as :func:`~.cond`,
+:func:`~.for_loop`, and :func:`~.while_loop`, that work with native PennyLane
+quantum operations. However, it can sometimes take a bit of work to rewrite
+existing Python code using these specific control flow functions. An experimental
+feature of PennyLane capture, AutoGraph, instead allows Pennylane capture to work
+with **native Python control flow**, such as if statements and for loops.
+
+Here, we'll aim to provide an overview of AutoGraph, as well as various
+restrictions and constraints you may discover.
+
+.. note::
+
+    When converting code in these examples, we will use the :func:`~.autograph.make_plxpr` function,
+    which uses AutoGraph by default.
+
+    When creating the initial plxpr representation, we must call the constructor function produced
+    by :func:`~.autograph.make_plxpr` with some initial values, which should have the same type and
+    shape as the values we intend to use when evaluating:
+
+    .. code-block:: python
+
+        from pennylane.capture import make_plxpr
+
+        def f(x):
+            if x > 5:
+                x = x ** 2
+            return x
+
+
+    >>> plxpr = make_plxpr(f)(0.0)  # x will be a float
+
+    Once the plxpr representation is created, we can evaluate it using
+
+    >>> from jax.core import eval_jaxpr
+    >>> eval_jaxpr(plxpr.jaxpr, plxpr.consts, 5.3)  # evaluate f(5.3)
+    [Array(28.09, dtype=float64, weak_type=True)]
+
+
+
+Using AutoGraph
+---------------
+
+The AutoGraph feature in PennyLane is supported by the ``diastatic-malt`` package, a standalone
+fork of the AutoGraph module in TensorFlow (
+`official documentation <https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/autograph/g3doc/reference/index.md>`_
+).
+
+The :func:`~pennylane.capture.make_plxpr` function uses AutoGraph by default. Consider a function using
+Python control flow:
+
+.. code-block:: python
+
+    dev = qml.device("default.qubit", wires=4)
+
+    @qml.qnode(dev)
+    def cost(weights, data):
+
+        for w in dev.wires.labels:
```

Comment:
@astralcai we have plans to support this, but because `Wires` don't present themselves as arrays, and the structure is not captured and we unroll the loop at compile time (at least this is how it works in Catalyst)

Reply:
In that case it might be a good idea to include a comment here about this. @lillian542

Reply:
Good point - in PennyLane, it raises an AutoGraph error (the catalyst fallback behaviour relies on queueing, which we are avoiding introducing to capture in PennyLane). 

I can also make the error message more informative, since this isn't very obvious.

Reply:
So actually this seems like it can just be fixed in a couple of lines in `Wires` by adding a `__jax_array__` property that returns an array of wire labels. Should I just fix it instead?

Reply:
Updated to remove `.labels`. It technically doesn't work on master yet, but it will once #6699 is merged.